src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,883 @@
+
+            public TableLeaseManager(
+                TableClient table,
+                AzureStorageOrchestrationService service,
+                AzureStorageOrchestrationServiceSettings settings,
+                string storageAccountName,
+                LeaseCollectionBalancerOptions options)
+            {
+                this.partitionTable = table;
+                this.service = service;
+                this.settings = settings;
+                this.storageAccountName = storageAccountName;
+                this.workerName = this.settings.WorkerId;
+                this.backgroundDrainTasks = new Dictionary<string, Task>();
+                this.options = options;
+            }
+
+            /// <summary>
+            /// Reads the partition table to determine the tasks the worker should do. Used by the PartitionManagerLoop.
+            /// During the iteration, the worker will first claim any available partitions with method `TryClaimLease`. 
+            /// Subsequently, if the partition is owned by this worker, it will proceed with method `CheckOwnershipLease`. 
+            /// However, if the partition is owned by other workers, it will utilize method `CheckOtherWorkerLease`.
+            /// If the shutdown is requested, then stop regular claim and balance process, and call method `TryDrainAndReleaseAllPartitions` to release all ownership leases.
+            /// </summary>
+            /// <param name="isShuttingDown">Indicates that the partition manager is in the process of shutting down.</param>
+            /// <param name="forcefulShutdownToken">Cancellation of this token forcefully aborts the partition manager loop.</param>
+            /// <returns> The <see cref="ReadTableReponse"/> incidates whether the worker is waiting to claim a stolen lease from other workers or working on releasing any ownership leases.</returns>
+            /// <exception cref="RequestFailedException">will be thrown if failed to update the partition table. Partition Manager loop will catch it and re-read the table to get the latest information.</exception>
+            public async Task<ReadTableReponse> ReadAndWriteTableAsync(bool isShuttingDown, CancellationToken forcefulShutdownToken)
+            {
+                var response = new ReadTableReponse();
+
+                IReadOnlyList<TableLease> partitions = this.partitionTable.Query<TableLease>(cancellationToken: forcefulShutdownToken).ToList();
+                var partitionDistribution = new Dictionary<string, List<TableLease>>(); 
+                int ownershipLeaseCount = 0;
+
+                foreach (TableLease partition in partitions)
+                {
+                    // In certain unhealthy situations, a worker may lose the lease without realizing it and continue listening
+                    // for queue messages. We check for that case here and stop dequeuing messages if we discover that
+                    // another worker currently owns the lease.
```

Comment:
nit: small phrasing suggestion
```suggestion
                    // In a worker becomes unhealthy, it may lose a lease without realizing it and continue listening
                    // for messages. We check for that case here and stop dequeuing messages if we discover that
                    // another worker currently owns the lease.
```